### ISSUE TITLE
New Python/C++ interface

### ISSUE THREAD

**Issue by wangkuiyi** [PaddlePaddle/Paddle#2750]
I read and followed this article http://intermediate-and-advanced-software-carpentry.readthedocs.io/en/latest/c++-wrapping.html, which compares the following interfacing technology:

1. manual wrapping. I followed this official Python document for more details: https://docs.python.org/2/extending/extending.html for some example programs.  There includes some complex boilerplate code -- parsing argument in each C function, build and return Python object in each C function, and the method list.

1. SWIG. It seems a general method that can generate bindings for various client languages, but not "native" enough.  Also, it takes some time to learn the interfacing language (*.i files).

1. ctypes.  This requires us to respecify the return type and other meta-data about each C function at the Python side, again.

1. SIP.  This is the Qt community's version of SWIG. We also need to learn an interfacing language.

1. Boost.Python.  This is some C++ templates that simplify the manual wrapping.  We no longer need to write a C wrapper function for each C++ function. Only a few extra lines in addition to the original C++ code are required to build a .so that can be called from Python.

I personally prefer Boost.Python.  Here is an example for your reference:

Suppose that we already have C++ functions like:

```cpp
char const* greet() {
   return "hello, world";
}
```

only the following few lines is required to build the Python-callable .so file:

```cpp
#include <boost/python.hpp>

BOOST_PYTHON_MODULE(hello_ext) {
  boost::python::def("greet", greet);
}
```

**Comment by Superjom** [comment]
Poost.python is simpler than swig, but it seems that it only supports python while SWIG supports more languages. 

For sure, we may only support Python in a long time, considering time and workload, but some choices should be made:

- if use `boost.python`, it is hard to support multiple language wrappers.
- `boost.python` is a more native wrapper for Python compared to SWIG, so if we use `boost.python`, it is easier to make a python-first system, in other words, better Python APIs (something like defining customize op/layer in Python? ) than TF/caffe2.

In short, `boost.python` for creating better Python APIs like PyTorch while `SWIG` for multiple language wrappers.

**Comment by typhoonzero** [comment]
> Boost.Python. This is some C++ templates that simplify the manual wrapping. We no longer need to write a C wrapper function for each C++ function. Only a few extra lines in addition to the original C++ code are required to build a .so that can be called from Python.

1. We need muli-language wrappers, at least Python and Go, because we'll need to call `Tensor`s and `Op`s in pserver side to do remote parameter optimization.
1. A c wrapper is also needed for high-performance online inference.

So I think we can't avoid making a c wrapper at last, once we have it, implement python extension is simple.

**Comment by wangkuiyi** [comment]
What is supposed to be included in the Go API?

I my mind, the Go API differs from the Python API significantly.  In particular, we don't need Go packages like paddle.op, paddle.layer, paddle.scope, paddle.variable. Is this correct?  @typhoonzero

**Comment by jacquesqiao** [comment]
There are another lib `https://github.com/pybind/pybind11` that just work like Boost.python, but is much more lightweight, If we will consider Boost.python, we can also have a look at this lib.

**Comment by typhoonzero** [comment]
> What is supposed to be included in the Go API?

At least paddle.variable or paddle.tensor I think. The current implementation of go pserver and optimizer use an independent implementation of tensor, at `paddle/optimizer/tensor.h`, better to use the new implement so we don't have two "tensor"s in the code base.

We don't need Go packages like paddle.op, paddle.layer, paddle.scope, paddle.variable indeed. Making parameter server as an "op" like tensorflow isn't what we intended to.

**Comment by dzhwinter** [comment]
consider that we just write operator in C++ and generate them in python. We cannot find a proper language binding generator library/generate technical.  Maybe it is too hard to generate OP for Go at present. 
I think we can just build a core system in c++, strong binding with python. other languages invoke functions from C-API binding.
Agree with @jacquesqiao, according to pybind11's doc [pybind11](https://github.com/pybind/pybind11) `pybind11 Similar to Boost.Python, but with a lean header-only implementation for C++11-capable compilers.`  It may be a better choice if we only consider the python c++ binding things. 

`ctypes` is not a good choice. Not only we need to write every function again in python side, but also it makes python binding tedious to maintain/upgrade.  In my mind, [mxnet](https://github.com/dmlc/mxnet/tree/master/python/mxnet), choose `ctypes` in the very beginning, but they maintain another logic in `cython` nowadays.

**Comment by reyoung** [comment]
If we want to use `C-API` to expose Paddle, which can make Paddle support various programming languages, `Cython` is a good choice. Because `Cython` is extremely easy to include `C` libraries.

Like:

```cython
cdef extern from "math.h":
    double sin(double x)
```

**Comment by reyoung** [comment]
Also, `pybind11` and `boost::Python` has a very major defect. It enforces the compiling Python version and running Python version EXACTLY SAME. It means if Paddle is compiled with `Python 2.7.2` but run with `Python 2.7.3`, an error will be raised.

**Comment by reyoung** [comment]
@wangkuiyi and all,

I write two demos, one used `pybind11`, other used `Cython+C-API`. They are:

* [PyBind11](https://github.com/PaddlePaddle/Paddle/pull/2793)
* [Cython+C-API](https://github.com/PaddlePaddle/Paddle/pull/2759)

The conclusion is:

* If we don't want a C-API, `PyBind11` is simpler. I think `PyBind11` has a great design of interface personally. I barely do not have to look at its documents to develop this demo. But there is one thing need to be careful, the [ownership](http://pybind11.readthedocs.io/en/stable/advanced/functions.html#return-value-policies) of C++ object. 

* If we have a C-API, use `Cython` is extremely easy. Just include the header and write some interface code.  `Cython` has a great interface, too, however, developing a C-API is noising.

**Comment by reyoung** [comment]
Fixed by #2793
